### ISSUE TITLE
[BUG] Can't send and receive video at the same time on Web <-> Unity setup

### ISSUE THREAD

**Issue by Kawottex** [Unity-Technologies/com.unity.webrtc#341]
Hello,
I already have a WebRTC implementation in Web/JavaScript that works well and I am currently implementing a Unity version with your library. I'm implementing a simple setup: web user send and receive audio/video and unity user send and receive audio/video.

Currently, my Unity implementation can send video to web, can receive video from web, but can't send and receive at same time.
To describe more, if the web user doesn't have a webcam (so, won't send video), Unity will be able to send video to web. Buf, if web does have a webcam (and will send video), Unity won't be able to send video to web.

I've looked at SDP messages and I don't see huge differences between both situations.
I've got two guesses, maybe it could help you:

- First, maybe Unity-WebRTC can only use one VideoStreamTrack at a time and use it only to send OR to receive, but not both. But it's strange, because my setup seems ok and it seems that I use one VideoStreamTrack to send and another to receive.
- Second, it may be because of codecs (seems that only VP8/VP9 can only decode), but I still don't understand why it can receive video when it doesn't send + I've checked the codecs used by the web part, and it seems to use VP8/VP9/H.264... (and if I force it to not use H.264, I have more problems). Moreover, my issue looks like this one #249 and I have the same problems if I use EcoderType.Hardware that the ones mentioned in the issue, but it seems that the problem in the issue is solved by using EncoderType.Software (which I did) + I don't have any errors on both sides.

Here is a simple representation of what I do to setup my peer connection and the receive of audio/video
```C#
// Init Unity.WebRTC library
Unity.WebRTC.WebRTC.Initialize(EncoderType.Software);
StartCoroutine(Unity.WebRTC.WebRTC.Update());

// Setup video and audio that will be sent
localVideoMat = new Material(Shader.Find("Diffuse"));
localVideoTex = new WebCamTexture();
localVideoMat.mainTexture = localVideoTex;
localVideoTex.Play();
localAudioStream = Audio.CaptureStream();

// Create peer connection and setup VideoStreamTrack and AudioStreamTrack
peerConnection = new RTCPeerConnection(ref rtcConfig);
peerConnection.OnTrack = OnTrack;
VideoStreamTrack videoSenderTrack = new VideoStreamTrack("video-sender", localVideoMat.mainTexture);
RTCRtpSender videoSender = peerConnection.AddTrack(videoSenderTrack);
rtcSenders.Add(videoSender);
foreach (AudioStreamTrack audioTrack in localAudioStream.GetAudioTracks())
{
    RTCRtpSender audioSender = peerConnection.AddTrack(audioTrack);
    rtcSenders.Add(audioSender);
}

// OnTrack, add the received VideoStreamTrack in a material applied on a gameObject videoPlayer
private void OnTrack(RTCTrackEvent _event)
{
   if (_event.Track.Kind == TrackKind.Video)
   {
       VideoStreamTrack videoTrack = (VideoStreamTrack)_event.Track;
       Texture videoMat = videoTrack.InitializeReceiver(1280, 720);
       videoPlayer.GetComponent<Renderer>().material.mainTexture = videoMat;
    }
}
```

I didn't show all the setup (ICE / Offer / Answer negociation) or all the verifications because I don't think it will be useful (+ I don't have any error that pop on both sides).
If you need more informations, don't hesitate to ask me.
**Edit**: It works on a Unity <-> Unity setup, so the problem is linked to the way Web and Unity implementations communicate

**Environment:**
 - OS: Windows
 - Browser: Firefox, Chrome
 - Unity Version: 2020.1.17f1
 - Package version M89 source code

Thank you,
Kevin

**Comment by Kawottex** [comment]
Edit: I've tried with a setup Unity-Unity and there's no problem. So it seems to be the differences between my web implementation and my Unity one (even if both are pretty similar). Do you have any idea on why could cause this difference ?

**Comment by gtk2k** [comment]
I can send and receive at the same time.
https://github.com/gtk2k/UnityWebRTC-UnityBrowser-Demo
(After doing Unity side, open the Web~ folder page and click the start button of the web page)

**Comment by Kawottex** [comment]
Thank you very much @gtk2k, your example helped me figuring out the problem. I don't know why, but the issue appears when I do 
```C#
RTCRtpSender audioSender = peerConnection.AddTrack(audioTrack);
```
In order to fix this, I had to do a `peerConnection.AddTransceiver(TrackKind.Audio);` in addition to the AddTrack. I don't understand why I have to do this, because I was thinking that having "offerToReceiveAudio = true" in my RTCOfferOptions permitted to avoid having to setup a new transceiver. Moreover, I don't have to add a transceiver for Video (and if I do, the bug come back) and I do the same thing for video as for audio (get a MediaTrack and set it with AddTrack to the PeerConnection).

I still have to test if the audio works well on both sides (can't test it right now), but even if my problem seems to be solved, I'd like to not close this issue until confirmation that this is the correct way to fix this.

**Comment by karasusan** [comment]
Bidirectional sample might be useful to fix your problem.
We have checked the combination of browser <-> Unity works fine.
https://docs.unity3d.com/Packages/com.unity.renderstreaming@3.0/manual/sample-bidirectional.html

**Comment by Kawottex** [comment]
> 
> 
> Bidirectional sample might be useful to fix your problem.
> We have checked the combination of browser <-> Unity works fine.
> https://docs.unity3d.com/Packages/com.unity.renderstreaming@3.0/manual/sample-bidirectional.html

Hello, thank for your answer, but, correct me if i'm wrong, it seems that in your example you are not using audio communication (thus, no AddTrack(audioTrack). It seems that my problem appears only when I try to implement audio sending.

**Comment by karasusan** [comment]
@Kawottex
Certainly. Audio receiver in Unity is not supported yet. 

Same issue.
https://github.com/Unity-Technologies/com.unity.webrtc/issues/296

**Comment by Kawottex** [comment]
> 
> 
> @Kawottex
> Certainly. Audio receiver in Unity is not supported yet.
> 
> Same issue.
> #296

Is it possible to implement audio sending only and then receive it in another WebRTC implementation ? (for example, Unity send audio + video to Web but only receive video)

**Comment by rollersteaam** [comment]
I don't know if this will help you (I may be misunderstanding what you're saying), but I've also had issues both displaying and sending a user's video (such as from a `WebCamTexture` in a self view). Attempting to do so causes the user's video to not show on both ends.

Solution goes roughly like this:

```csharp
webcam = new WebCamTexture(1280, 720);
webcam.Play();

var gfxType = SystemInfo.graphicsDeviceType;
var format = Unity.WebRTC.WebRTC.GetSupportedGraphicsFormat(gfxType);

// Create a duplicate texture to allow both sending and playback
webcamRt = new RenderTexture(1280, 720, 0, Unity.WebRTC.WebRTC.GetSupportedRenderTextureFormat(gfxType));

videoStream = new MediaStream();
var track = new VideoStreamTrack("uservideo", webcamRt);

videoStream.AddTrack(track);

peerConnection.AddTrack(track);
```

Then in an Update method do this:

```csharp
if (webcam != null && webcamRt != null)
{
    Graphics.Blit(webcam, webcamRt);
}
```

Then when displaying the video, make sure to not use `webcamRt` but instead use the `webcam` texture. I guess this is because `webcamRt`'s ownership is given to the WebRTC library and also reading from it causes problems.

```csharp
videoSurface.material.mainTexture = webcam;
```

**Comment by karasusan** [comment]
@Kawottex 
We released the latest version 2.4.0-exp.4 that containing audio receiver support.
Let you try it.

**Comment by Kawottex** [comment]
Hello, thank you for your work ! I implemented audio support on my application and it works well to send audio but I've got some issues when receiving. In fact, on the OnTrack callback, I correctly receive an AudioStreamTrack but the Source and Renderer are all null. I'll try to investigate and if I can't find out, I'll open a new ticket. But this one is closed, thanks.
